File: domain/src/main/java/CommandHandlerBase.java
import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Future;

public abstract class CommandHandlerBase<T extends DomainCommand> implements IHandleCommands<T> {
    private final Repository _repository;
    private final IValidateTokens _tokenValidator;

    protected CommandHandlerBase(Repository repository, IValidateTokens tokenValidator) {
        _repository = repository;
        _tokenValidator = tokenValidator;
    }

    public Future<CommandResponse> Handle(T command, MessageHeaders headers) {
        final String userToken = _tokenValidator.validate(headers.getUserToken());
        try {
            if (!verifyUserToken(userToken).get()) {
                return new CompletableFuture<CommandResponse>().completeAsync(() -> new CommandResponse(command.getAggregateId(), command.getVersion(), null, command.getCorrelationId()));
            }
        } catch (Exception e) {
            return new CompletableFuture<CommandResponse>().completeAsync(() -> new CommandResponse(command.getAggregateId(), command.getVersion(), e.getMessage(), command.getCorrelationId()));
        }

        try {
            return handleInternal(command, headers).thenApply(r -> Objects.requireNonNullElseGet(r, () -> new CommandResponse(command.getAggregateId(), command.getVersion(), null, command.getCorrelationId())));
        } catch (Exception ex) {
            return new CompletableFuture<CommandResponse>().completeAsync(() -> new CommandResponse(command.getAggregateId(), command.getVersion(), ex.getMessage(), command.getCorrelationId()));
        }
    }

    protected CompletableFuture<Boolean> verifyUserToken(String token) {
        return new CompletableFuture<Boolean>().completeAsync(() -> true);
    }

    protected abstract CompletableFuture<CommandResponse> handleInternal(T command, MessageHeaders headers);

    protected <TAggregate extends Aggregate> Future<TAggregate> get(String id) {
        return _repository.getById(id);
    }

    protected <TAggregate extends Aggregate> Future<Boolean> save(TAggregate aggregate) {
        return _repository.save(aggregate, h -> {
        });
    }
}

File: modules/domain/src/main/kotlin/org/openmedstack/domain/CommandHandlerBase.kt
package org.openmedstack.domain

import org.openmedstack.IValidateTokens
import org.openmedstack.MessageHeaders
import org.openmedstack.commands.CommandResponse
import org.openmedstack.commands.DomainCommand
import org.openmedstack.commands.IHandleCommands
import org.openmedstack.eventstore.DuplicateCommitException
import org.openmedstack.eventstore.Repository
import org.openmedstack.eventstore.Snapshot
import java.util.concurrent.CompletableFuture

abstract class CommandHandlerBase protected constructor(private val _repository: Repository, private val _tokenValidator: IValidateTokens) : IHandleCommands {
    override fun handle(command: DomainCommand, messageHeaders: MessageHeaders): CompletableFuture<CommandResponse> {
        return _tokenValidator.validate(messageHeaders.userToken)
            .thenComposeAsync { userToken: String? -> createResponse(userToken, command, messageHeaders) }
    }

    private fun createResponse(
        userToken: String?,
        command: DomainCommand,
        headers: MessageHeaders?
    ): CompletableFuture<CommandResponse?> {
        return verifyUserToken(userToken).thenComposeAsync { b: Boolean ->
            if (b) handleInternal(command, headers)
            else CompletableFuture.completedFuture(CommandResponse.error(command, "Invalid token"))
        }
    }

    protected fun verifyUserToken(token: String?): CompletableFuture<Boolean> {
        return CompletableFuture.supplyAsync { true }
    }

    protected abstract fun handleInternal(command: DomainCommand, headers: MessageHeaders?): CompletableFuture<CommandResponse>

    protected operator fun <TAggregate : Aggregate, TMemento : Memento> get(
        type: Class<TAggregate>,
        mementoType: Class<Snapshot<TMemento>>,
        id: String
    ): CompletableFuture<TAggregate> {
        return _repository.getById(type, mementoType, id)
    }

    @Throws(DuplicateCommitException::class)
    protected fun <TAggregate : Aggregate, TMemento : Memento> save(aggregate: TAggregate): CompletableFuture<Boolean> {
        return _repository.save<TAggregate, TMemento>(aggregate) { }
    }
}
